force-app/main/default/classes/SOQL.cls

```diff
@@ -288,21 +418,34 @@ public with sharing class SOQL {
      */
     public static List<Account> findAccountsByLeadEmailDomains(Set<Id> leadIds) {
         // Query the Leads
-
-        // Create a set of email domains from the Leads
-
-        // Loop through each Lead
-
-            // Get the email from the Lead
-
-            // Check if the email is not null
-
-                // Add the email domain to the set
-
-        
-        // Query the Accounts
-        return null; // Replace null with the variable you used to store the result
+        List<Lead> leadList = [
+            SELECT Id, Email
+            FROM Lead
+            WHERE Id IN :leadIds WITH SECURITY_ENFORCED
+        ];
+    
+        Set<String> emailDomains = new Set<String>();
+        for (Lead lead : leadList) {
+            if (lead.Email != null && lead.Email.indexOf('@') != -1) {
+                emailDomains.add(lead.Email.split('@')[1]);
+            }
+        }
+    
+        String query = 'SELECT Id, Website FROM Account WHERE';
```

Comment:
You did well implementing dynamic SOQL, but its not necessary to solve this method. You can also use a direct query and that might simplify things